.ci/convert_notebooks.py

```diff
@@ -0,0 +1,88 @@
+                        help="timeout for notebook execution")
+    parser.add_argument("--rst_dir", type=Path,
+                        help="rst files output directory", default=Path("rst"))
+    parser.add_argument("--html_dir", type=Path,
+                        help="html files output directory", default=Path("html"))
+    parser.add_argument("--markdown_dir", type=Path,
+                        help="markdown files output directory", default=Path("markdown"))
+
+    return parser.parse_args()
+
+
+def prepare_ignore_execution_list(input_file):
+    with Path(input_file).open("r") as f:
+        lines = f.readlines()
+    return list(map(str.strip, lines))
+
+
+def main():
+    args = arguments()
+    ignore_execution_list = []
+    failed_notebooks = []
+    markdown_failed = []
+    rst_failed = []
+    html_failed = []
+    if args.exclude_execution_file is not None:
+        ignore_execution_list = prepare_ignore_execution_list(
+            args.exclude_execution_file)
+    ROOT = Path(__file__).parents[1]
+    notebooks_dir = ROOT / "notebooks"
+    notebooks = sorted(list(notebooks_dir.rglob('**/*.ipynb')))
+    for notebook in notebooks:
+        notebook_path = notebook.relative_to(ROOT)
+        notebook_executed = notebook_path.parent / notebook_path.name.replace(".ipynb", "-with-output.ipynb")
+        start = time.perf_counter()
+        print(f"Convert {notebook_path}")
+        if str(notebook_path) not in ignore_execution_list:
+            retcode = subprocess.run(["jupyter", "nbconvert",  "--log-level=INFO", "--execute", "--to",  "notebook", "--output",
+                                     str(notebook_executed),  '--output-dir', str(ROOT), '--ExecutePreprocessor.kernel_name=python3', str(notebook_path)], timeout=args.timeout).returncode
+            if retcode:
+                failed_notebooks.append(str(notebook_path))
+                continue
+        else:
+            shutil.copyfile(notebook_path, notebook_executed)
+        markdown_retcode = subprocess.run(["jupyter", "nbconvert", "--to", "markdown", str(notebook_executed), "--output-dir", str(args.markdown_dir),
+                                          "--TagRemovePreprocessor.remove_all_outputs_tags=hide_output --TagRemovePreprocessor.enabled=True"], timeout=args.timeout).returncode
+
+        html_retcode = subprocess.run(["jupyter", "nbconvert", "--to", "html", str(notebook_executed), "--output-dir", str(args.html_dir),
+                                          "--TagRemovePreprocessor.remove_all_outputs_tags=hide_output --TagRemovePreprocessor.enabled=True"], timeout=args.timeout).returncode
+        rst_retcode = subprocess.run(["jupyter", "nbconvert", "--to", "rst", str(notebook_executed), "--output-dir", str(args.rst_dir),
+                                          "--TagRemovePreprocessor.remove_all_outputs_tags=hide_output --TagRemovePreprocessor.enabled=True"], timeout=args.timeout).returncode
```

Comment:
Shouldn't be under `if str(notebook_path) not in ignore_execution_list`?

Reply:
no, we have some notebooks which store output inside and we no need to spend time to execute them on ci (e.g. notebook requires GPU, but ci machine does not have it), so all actions except execution are acceptable for them, we just avoid execution